### ISSUE TITLE
Provide a more flexible way to define what is a failure

### ISSUE THREAD

**Issue by oxsdd** [vert-x3/vertx-circuit-breaker#65]
#### Describe the feature

We suggest enhancing the API so that the criteria for "failure" is more flexible than the success/failure of the `Promise`, perhaps similar to Failsafe: https://failsafe.dev/policies/#failure-handling

#### Use cases

When using the Circuit Breaker for either `EventBus` request-reply, or a Service Proxy, the reply may be an "application level" error that the CircuitBreaker should not consider a failure, and return to the caller to handle. Indeed  `ServiceException` provides a `failureCode` presumably for that purpose (that is how we use it).

For example, a Service Proxy service  might return an application-level "not found" error by replying with a `ServiceException`. In this case, the `CircuitBreaker` should not consider this a failure. We probably only want the `CircuitBreaker` to consider a `ServiceException` to be a failure when `failureType` is either `ReplyFailure.TIMEOUT` or `ReplyFailure.NO_HANDLERS`. For example, perhaps using a `Predicate` to indicate when the operation has failed:

```java
Predicate<AsyncResult<?>> isFailed = ar -> {
  if (ar.failed() && ar.cause() instanceof ServiceException) {
    ReplyFailure rf = ((ServiceException)ar.cause()).failureType();
    return rf == ReplyFailure.TIMEOUT || rf == ReplyFailure.NO_HANDLERS;
  }
  return false;
};

CircuitBreakerOptions options = new CircuitBreakerOptions().setFailurePredicate(isFailed);
CircuitBreaker breaker = CircuitBreaker.create("test", vertx, options);

// Predicate from options:
breaker.execute( promise -> myServiceProxy.myOperation("arg1", promise)).onComplete(ar -> {// .... });

// or specify failure predicate on each call
breaker.execute( promise -> myServiceProxy.myOperation("arg1", promise), isFailed).onComplete(ar -> { //.... });
```

If the failure predicate is not specified, the default is simply `ar -> ar.failed()`, which is equivalent to the current API.

The main value in this is that it makes using the `CircuitBreaker` with Service Proxy services, and `EventBus` request-reply easier.  That is significant  since the `EventBus` is a best-effort delivery system and thus often requires retries to handle system failures and even routine system downtime for maintenance.

#### Contribution

We may be able to volunteer to provide this feature at some point in the future, but currently do not have the resources. We have already handled this situation in our applications, but  the solution is not very clean.

**Comment by tsegismont** [comment]
Sounds interesting. Would you like to contribute?

**Comment by oxsdd** [comment]
I would like to but it will be a couple of months until time is available. This would be my first contribution and since it would involve changes to the API, there would be some ramp-up time to make sure we got it right. So please consider me a potential contributor if no others are available before then. I would definitely be available to test development snapshots if others wish to contribute before then. Thanks!

**Comment by oxsdd** [comment]
Sorry for the delayed update. Unfortunately we won't be able to contribute, so feel free to close if that's appropriate.

We have moved to using [Failsafe](https://failsafe.dev/) in some cases, since it supports Vert.x, it has the additional flexibility that we require, and we can use it in non-Vert.x projects.

This is in no way a criticism of the Vert.x CircuitBreaker. It has some nice features like EventBus integration that are quite useful.  We will continue to use it in some cases. But it's definition of *failure* being  the simple success or failure of the `Future` is too limited in some of our use cases.

See [this link](https://github.com/failsafe-lib/failsafe/blob/master/examples/src/main/java/dev/failsafe/examples/VertxExample.java) for how Failsafe integrates with Vert.x

Thanks!

**Comment by CheesyBoy123** [comment]
@tsegismont I saw this was a help wanted issue, I've gone ahead and given this a try in https://github.com/vert-x3/vertx-circuit-breaker/pull/72/files let me know what you think, thanks!

**Comment by tsegismont** [comment]
Thank you @CheesyBoy123 , I'll review it

**Comment by tsegismont** [comment]
Closed by #72 72
